### DISCUSSION TITLE
fetch RequestDuplex is "full" even set as "half"

### DISCUSSION THREAD

**Discussion by zizifn** [nodejs/undici#1760] [Q&A]
fetch spec say,
>[duplex](https://fetch.spec.whatwg.org/#dom-requestinit-duplex)
"half" is the only valid value and it is for initiating a half-duplex fetch (i.e., the user agent sends the entire request before processing the response). "full" is reserved for future use, for initiating a full-duplex fetch (i.e., the user agent can process the response before sending the entire request). This member needs to be set when [body](https://fetch.spec.whatwg.org/#dom-requestinit-body) is a [ReadableStream](https://streams.spec.whatwg.org/#readablestream) object. See [issue #1254](https://github.com/whatwg/fetch/issues/1254) for defining "full".

But seems to fetch from undici always use `full`, so does this is **expected** in undici / nodejs ?

Below is my test case in 
>nodejs v16.18.0
>undici 5.12.0

Nodejs http server
``` javascript
import http from "node:http";

const requestListener = async function (req, res) {
  console.log("start");
  if (req.method === "POST") {
    let index = 0;
    res.writeHead(200);
    for await (const chunk of req) {
      console.log(`client send ${chunk.toString()}`);
      res.write(
        `server response ${index++}, after client send ${chunk.toString()}`
      );
    }
    res.end();
  } else {
    res.setHeader("Allow", "POST");
    res.writeHead(405).end("Method Not Allowed");
  }
};

const server = http.createServer(requestListener);
server.listen(8080);

```

Below is the fetch client
```javascript
import {
  ReadableStream,
  TextDecoderStream,
  TextEncoderStream,
} from "node:stream/web";
import { fetch } from "undici";

try {
  const readableStream = new ReadableStream({
    index: 0,
    start(controller) {
      const interval = setInterval(() => {
        controller.enqueue(`client send ${this.index++}`);
      }, 500);

      setTimeout(() => {
        clearInterval(interval);
        controller.close();
      }, 10_000);
    },
  }).pipeThrough(new TextEncoderStream());

  const url = "http://localhost:8080";

  const resp = await fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "text/plain",
    },
    body: readableStream,
    duplex: "half",
  });

  for await (let chunk of resp.body.pipeThrough(new TextDecoderStream())) {
    console.log(chunk);
  }
} catch (err) {
  console.log(err);
}

```
Screen the test result
![fetch-duplex](https://user-images.githubusercontent.com/1803942/200370762-69adb20c-b854-4f7e-9a7f-be987d19d536.gif)

**Comment by ronag** [comment]
@KhafraDev I don't think undici currently even supports half duplex.

**Reply by zizifn** [reply]
> @KhafraDev I don't think undici currently even supports half duplex.

from my above test code, fetch in undici is support full duplex if I understand things correct.

**Reply by zizifn** [reply]
@ronag  update screen result about above code.

**Reply by KhafraDev** [reply]
Yeah, but the spec tells us to return `half` for the duplex option; I guess it's not meant to support full duplex *just* yet. Fetch should probably support half duplex and maybe go against the spec and add in a full option that maintains the current behavior?

**Reply by zizifn** [reply]
I like `full`, and spec also allows "full" for future use. Just want to confirm does this is expected in the future as nodejs will also have native fetch. 
But this is totally deponed on you or nodejs team, if you or team decide to use `full`, please update doc, or I can help update doc?

**Reply by KhafraDev** [reply]
A doc change would be great for now, stating the the "duplex" option doesn't affect behavior currently & fetch requests are currently full duplex.

**Reply by zizifn** [reply]
PR for update doc... https://github.com/nodejs/undici/pull/1765
